### PR TITLE
Fix O(labels * events) blowup in get_task_seq_bounds_and_labels (#92)

### LINKED ISSUE #92: get_task_seq_bounds_and_labels materializes O(labels × events) intermediate; overflows u32 index / OOMs on skewed label distributions
### Summary

`MEDSPytorchDataset.get_task_seq_bounds_and_labels` ([`pytorch_dataset.py` L76](https://github.com/mmcdermott/meds-torch-data/blob/main/src/meds_torchdata/pytorch_dataset.py#L76)) computes per-label `end_idx` by:

1. joining `label_df` against `schema_df` (whose `time` column is a list of event timestamps per subject),
2. `.explode(time_name)` — producing a flat `label × event` rowset,
3. `group_by` + `search_sorted(...).last()` to recover one row per label.

This materializes an intermediate of size `Σ_subject (labels_for_subject × events_for_subject)` to compute what is logically a per-label binary search — O((labels + events) · log events) total.

### Symptom

On realistic MEDS data with a heavy-tailed per-subject event count and labels sampled proportionally to events (so labels/subject ∝ events/subject), the intermediate row count scales like `Σ events²` per subject. Hits:

```
thread 'polars-0' panicked at crates/polars-core/src/chunked_array/ops/chunkops.rs:152:13:
Polars' maximum length reached. Consider installing 'polars-u64-idx'.
```

Swapping to `polars-u64-idx` gets past the u32 cap but then OOMs on the same code path at the `.explode(time_name)` step.

### Repro sketch

Any workload where labels-per-subject correlates with events-per-subject will surface this. Concretely: a sampler that draws prediction-time contexts with replacement from a flat events DF (so subjects with more events get more labels) on a shard where a few subjects have 10–100× the median event count.

Workaround on the caller side: hard-cap labels per subject before handing them to the datamodule. But that requires every caller to know this implementation detail.

### Expected

`end_idx = search_sorted(time_list_for_subject, prediction_time, side="right")` is a binary search; compute should be O(log events) per label, memory O(labels + events).

### Suggested fix

Replace the explode path with a `join_asof` against a flat events DF carrying a per-subject row index, e.g.:

```python
events_with_idx = events_df.with_columns(
    pl.int_range(0, pl.len()).over(DataSchema.subject_id_name).alias("_event_idx")
).sort(DataSchema.subject_id_name, DataSchema.time_name)

return (
    label_df.sort(DataSchema.subject_id_name, LabelSchema.prediction_time_name)
    .join_asof(
        events_with_idx,
        left_on=LabelSchema.prediction_time_name,
        right_on=DataSchema.time_name,
        by=DataSchema.subject_id_name,
        strategy="backward",
    )
    .with_columns((pl.col("_event_idx") + 1).alias(cls.END_IDX))
    .select(out_cols)
)
```

Same semantics (`end_idx` = count of events with `time <= prediction_time`), no cartesian intermediate. Requires having the flat events DF available at `schema_df` construction time — may involve plumbing changes.

### Environment

- `meds-torch-data` 0.6.3
- `polars` 1.30 / reproduces with `polars-u64-idx` 1.30 (same algorithm, OOM instead of overflow)

### PR DESCRIPTION
**Pull request by mmcdermott**
Closes #92.

## Summary

`MEDSPytorchDataset.get_task_seq_bounds_and_labels` joined `label_df` against `schema_df` (list-valued `time` column) and then `.explode()`ed, materializing an intermediate of size `sum_subject(labels * events)` before collapsing back to one row per label. On skewed cohorts (a few subjects carrying most of the labels *and* most of the events), this row count is `sum(events^2)` per subject and:

- overflows polars' default u32 row index (`Polars' maximum length reached` panic), or
- OOMs on the `.explode(time)` call itself if the u32 cap is lifted via `polars-u64-idx`.

This PR replaces the explode path with a flat events DF plus `join_asof(strategy="backward")`:

1. Explode `schema_df.time` once — intermediate is `O(total_events)`, not `O(labels * events)`.
2. Attach per-subject `_event_idx` via `int_range().over(subject_id)`.
3. `join_asof` each label against the flat events table on `prediction_time <= event_time` within `by=subject_id`. Linear merge, `O((labels + events) log events)` total.
4. `end_event_index = _event_idx + 1` with `null -> 0` for labels preceding all events for their subject.

Function signature unchanged (still takes list-valued `schema_df`) so no caller plumbing changes.

## Benchmark (scaled-down repro — full issue scale OOMs a dev box)

One skewed subject with 10k events x 500 labels + 99 unskewed (100 events, 1 label) subjects, 5M cartesian intermediate:

| impl           | wall time |
|----------------|-----------|
| current (dev)  | 0.196 s   |
| this PR        | 0.017 s   |
| **speedup**    | **~12x**  |

Asymptotic: `O(labels * events)` -> `O((labels + events) * log events)`. The full issue scale (multi-subject heavy tail with 50k+ events/subject and thousands of labels/subject) goes from "OOMs or u32-overflows" to "completes".

## Test plan

- [x] Existing doctest passes unchanged — covers the happy path (subjects with multiple prediction times + boolean labels, and the label-less variant).
- [x] New doctest covering the `prediction_time < first_event` edge case, where the previous `search_sorted` side-right returns 0 and the new `join_asof` null-fills to 0 via `.fill_null(0)`.
- [x] Full local test suite: 47 passed, 2 skipped (lightning).
- [ ] CI green.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/meds_torchdata/pytorch_dataset.py [hunk 1]
```diff
@@ -146,29 +146,75 @@ def get_task_seq_bounds_and_labels(cls, label_df: pl.DataFrame, schema_df: pl.Da
...
+        sid = DataSchema.subject_id_name
+        pt = LabelSchema.prediction_time_name
+        time_col = DataSchema.time_name
+
+        events_flat = (
+            schema_df.lazy()
+            .select(sid, time_col)
+            .explode(time_col)
+            .with_columns(pl.int_range(pl.len()).over(sid).alias("_event_idx"))
+            .sort(sid, time_col)
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/pytorch_dataset.py]: **Real bug, fixed in f64048f.** You're right — `int_range().over(sid)` attached a pre-sort index that then got detached by the `.sort(sid, time_col)`. Reordered the pipeline to sort first, then compute the index over the sorted stream:

```python
events_flat = (
    schema_df.lazy()
    .select(sid, time_col)
    .explode(time_col)
    .sort(sid, time_col)                                                    # sort FIRST
    .with_columns(pl.int_range(pl.len()).over(sid).alias("_event_idx"))    # index after
)
```

MEDS tokenization does produce sorted time lists upstream, but relying on that at the `join_asof` layer is fragile. Added `test_unsorted_schema_time_list` in `tests/test_get_task_seq_bounds_equivalence.py` that feeds deliberately shuffled per-subject time lists and asserts equivalence with the Python-loop reference — that test would have failed on the previous version of this PR.

File: src/meds_torchdata/pytorch_dataset.py [hunk 2]
```diff
@@ -146,29 +146,75 @@ def get_task_seq_bounds_and_labels(cls, label_df: pl.DataFrame, schema_df: pl.Da
...
 
         return (
-            label_df.join(schema_df, on=DataSchema.subject_id_name, how="inner", maintain_order="left")
+            label_df.lazy()
             .with_row_index("_row")
-            .explode(DataSchema.time_name)
-            .group_by(group_cols, maintain_order=True)
-            .agg(end_idx_expr)
+            .filter(pl.col(sid).is_in(subj_with_events))
+            .sort(sid, pt)
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/pytorch_dataset.py]: Agreed — swapped to a semi-join in f64048f. Replaces `subj_with_events = ...unique().to_series().implode()` + `filter(is_in(...))` with `.join(schema_df.lazy().select(sid).unique(), on=sid, how="semi")`. Stays lazy, reads closer to the intended inner-join semantics, and drops the list-broadcast codepath.

[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/pytorch_dataset.py:211]: Fixed in f64048f. You and ChatGPT both flagged this. The claim about subjects-absent-from-`schema_df` getting null indices has actually never been true in practice — both the old explode-and-group-by implementation (which used `how="inner"` on the initial join) and the new `join_asof`-based one drop those labels entirely. Updated the docstring to describe the actual inner-join behavior and added a parametrized `drops_absent_subject` case in the new equivalence test.

File: src/meds_torchdata/pytorch_dataset.py [hunk 3]
```diff
@@ -146,29 +148,80 @@ def get_task_seq_bounds_and_labels(cls, label_df: pl.DataFrame, schema_df: pl.Da
...
-
+        out_cols = [sid, cls.END_IDX, pt]
         if cls.LABEL_COL in label_df.collect_schema().names():
-            group_cols.append(cls.LABEL_COL)
             out_cols.append(cls.LABEL_COL)
 
+        # `join_asof` with `by` behaves like a left join (non-matching subjects get null
+        # on the right). We want inner-join semantics — labels for subjects absent from
+        # `schema_df` are dropped entirely — so semi-join the labels against the subjects
+        # that have events first. Keeps the whole thing lazy.
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/pytorch_dataset.py]: Good catch — fixed in 5e5ed5f. You're right: the semi-join is on `schema_df.select(sid).unique()`, which keeps subjects regardless of whether their `time` list is empty. An empty-time subject is still "present in schema_df" and survives this filter; the downstream `join_asof` then null-matches on the empty time column and `.fill_null(0)` gives those labels `end_idx = 0`. Reworded the comment to reflect that, and explicitly called out the empty-list case so the next reader doesn't misread it the same way. (The `empty_event_list` case in the @example anchor on the hypothesis test exercises exactly this path.)

File: src/meds_torchdata/pytorch_dataset.py [hunk 4]
```diff
@@ -146,29 +148,82 @@ def get_task_seq_bounds_and_labels(cls, label_df: pl.DataFrame, schema_df: pl.Da
             │ 3          ┆ 1               ┆ 2020-01-02 00:00:00 │
             │ 3          ┆ 1               ┆ 2020-01-03 00:00:00 │
             └────────────┴─────────────────┴─────────────────────┘
+
+            Prediction times at or before a subject's first event collapse to `end_idx = 0`
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/pytorch_dataset.py]: Good catch — fixed in 62a760c. "At or before" was wrong: for `prediction_time == times[0]` the count-of-events-with-`time <= prediction_time` is 1, not 0. The existing `subject 3 / 2020-01-01` row in the main doctest (`end_idx = 1`) already demonstrates this; only my *new* edge-case paragraph got the wording wrong. Reworded to "strictly before" with a parenthetical explicitly calling out the equality case so the contract is unambiguous.
